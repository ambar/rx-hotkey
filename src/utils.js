export const canUseDOM = typeof document !== 'undefined'

// https://mdn.io/KeyboardEvent/key/Key_Values#Modifier_keys
const isModifier = key => ['Control', 'Shift', 'Alt', 'Meta'].includes(key)

const isInput = el =>
  /^(input|textarea|select)$/i.test(el.tagName) || el.isContentEditable

export const isValidEvent = event =>
  !isInput(event.target) && !isModifier(event.key)

export const parseHotkeys = keys =>
  keys.split(' ').map(key => {
    const hotkey = {
      key: null,
      ctrlKey: false,
      altKey: false,
      metaKey: false,
      shiftKey: false,
    }
    return key
      .split('+')
      .map(k => k.toLowerCase())
      .reduce((o, k) => {
        if (k === 'ctrl') {
          o.ctrlKey = true
        } else if (k === 'alt') {
          o.altKey = true
        } else if (k === 'shift') {
          o.shiftKey = true
        } else if (k === 'meta') {
          o.metaKey = true
        } else {
          o.key = k
        }
        return o
      }, hotkey)
  })

const matchObject = (obj, partial) =>
  Object.keys(partial).every(k => partial[k] === obj[k])

const shiftedKeys = '~!@#$%^&*()_+{}|:"<>?'

export const matchEvent = (event, hotkey) =>
  matchObject(
    event,
    shiftedKeys.includes(hotkey.key) ? {...hotkey, shiftKey: true} : hotkey
  )
