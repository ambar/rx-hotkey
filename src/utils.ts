export const canUseDOM = typeof document !== 'undefined'

// https://mdn.io/KeyboardEvent/key/Key_Values#Modifier_keys
const isModifier = (key: string) =>
  ['Control', 'Shift', 'Alt', 'Meta'].includes(key)

const isInput = (el: HTMLElement) =>
  /^(input|textarea|select)$/i.test(el.tagName) || el.isContentEditable

export const isValidEvent = (event: KeyboardEvent) =>
  !isInput(event.target as HTMLElement) && !isModifier(event.key)

const modifiers = ['ctrl', 'alt', 'shift', 'meta'] as const
type Modifier = typeof modifiers[number]
type ComboKey = {key: string} & {
  [k in `${Modifier}Key`]: boolean
}

export const parseHotkeys = (keys: string) =>
  // eg. `ctrl+k f`
  keys.split(' ').map((key) => {
    // eg. `ctrl+k`
    const list = key.split('+').map((k) => k.toLowerCase())
    const parsed = {key: ''} as ComboKey
    modifiers.forEach((x) => {
      parsed[`${x}Key`] = list.includes(x)
    })
    list.some((x) => {
      if (!modifiers.includes(x as Modifier)) {
        parsed.key = x
        return true
      }
    })
    return parsed
  })

type PlainObject = Record<string, unknown>
export const matchObject = <T extends PlainObject>(
  obj: T,
  partial: T,
  comparator = (a: unknown, b: unknown) => a === b
) => Object.keys(partial).every((k) => comparator(partial[k], obj[k]))

const shiftedKeys = '~!@#$%^&*()_+{}|:"<>?'

export const matchEvent = (event: KeyboardEvent, hotkey: ComboKey) =>
  matchObject(
    event as unknown as PlainObject,
    shiftedKeys.includes(hotkey.key) ? {...hotkey, shiftKey: true} : hotkey,
    (a, b) =>
      typeof a === 'string' && typeof b === 'string'
        ? a.toLowerCase() === b.toLowerCase()
        : a === b
  )
