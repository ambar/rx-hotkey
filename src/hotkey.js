import {fromEvent, Subject} from 'rxjs'
import {
  take,
  filter,
  bufferCount,
  switchMap,
  startWith,
  timeoutWith,
} from 'rxjs/operators'
import {canUseDOM, isValidEvent, parseHotkeys, matchEvent} from './utils'

const maxSequenceTimeout = 1500
const keyBindings = new Map()
const keyBindings$ = new Subject()

export const createHotkey = (element, {scope} = {}) => {
  if (!canUseDOM) {
    return () => () => {}
  }

  const keys$ = fromEvent(element, 'keydown').pipe(filter(isValidEvent))
  const scopeId = scope || element

  return (keys, handler, options) => {
    let scopedKeyBindings = keyBindings.get(scopeId)
    if (!scopedKeyBindings) {
      scopedKeyBindings = new Map()
      keyBindings.set(scopeId, scopedKeyBindings)
    }
    scopedKeyBindings.set(keys, {...options})
    keyBindings$.next(new Map(keyBindings))

    const hotkeys = parseHotkeys(keys)
    const matchEvents = filter(events =>
      (Array.isArray(events) ? events : [events]).every((e, i) =>
        matchEvent(e, hotkeys[i])
      )
    )
    const output$ =
      hotkeys.length === 1
        ? keys$.pipe(matchEvents)
        : keys$.pipe(
            take(1),
            switchMap(first$ =>
              keys$.pipe(
                startWith(first$),
                bufferCount(hotkeys.length, 1),
                timeoutWith(maxSequenceTimeout, output$)
              )
            ),
            matchEvents
          )

    const subscription = output$.subscribe(handler)
    return () => {
      scopedKeyBindings.delete(keys)
      if (!scopedKeyBindings.size) {
        keyBindings.delete(scopeId)
      }
      keyBindings$.next(new Map(keyBindings))
      subscription.unsubscribe()
    }
  }
}

export const getKeyBindings = () => keyBindings
export const listenKeyBindings = handler => {
  const subscription = keyBindings$.subscribe(handler)
  return () => subscription.unsubscribe()
}

export const hotkey = createHotkey(canUseDOM ? document : undefined, {
  scope: 'global',
})
