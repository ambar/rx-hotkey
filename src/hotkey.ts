import {fromEvent, Subject, Observable} from 'rxjs'
import {
  take,
  filter,
  bufferCount,
  switchMap,
  startWith,
  timeoutWith,
} from 'rxjs/operators'
import {canUseDOM, isValidEvent, parseHotkeys, matchEvent} from './utils'

type Key = string
// {[key]: metadata}
type Bindings = Map<Key, object>
// {[scope]: {[key]: metadata}}
type ScopedBindings = Map<string, Bindings>
const maxSequenceTimeout = 1500
const keyBindings: Bindings = new Map()
const keyBindings$ = new Subject<ScopedBindings>()

type HotkeyOptions = {
  /** scope key in key bindings, defaults to `global` */
  scope?: string
}

const ensureArray = <T extends unknown>(x:  T): T[] => (Array.isArray(x) ? x : [x])

export const createHotkey = (element?: HTMLElement | Document, {scope}: HotkeyOptions = {}) => {
  if (!canUseDOM) {
    return () => () => {}
  }

  const keys$ = fromEvent<KeyboardEvent>(element!, 'keydown').pipe(filter(isValidEvent))
  const scopeId = scope || 'global'

  return (keys: Key, handler: (...e: KeyboardEvent[]) => void, metadata?: object) => {
    let scopedKeyBindings = keyBindings.get(scopeId) as Bindings
    if (!scopedKeyBindings) {
      scopedKeyBindings = new Map()
      keyBindings.set(scopeId, scopedKeyBindings)
    }
    scopedKeyBindings.set(keys, {...metadata})
    keyBindings$.next(new Map(keyBindings) as ScopedBindings)

    const hotkeys = parseHotkeys(keys)
    const matchEvents = filter((events: KeyboardEvent | KeyboardEvent[]) =>
      (ensureArray(events) as KeyboardEvent[]).every((e, i) =>
        matchEvent(e, hotkeys[i])
      )
    )
    const output$: Observable<KeyboardEvent | KeyboardEvent[]> =
      hotkeys.length === 1
        ? keys$.pipe(matchEvents)
        : keys$.pipe(
            take(1),
            switchMap(first$ =>
              keys$.pipe(
                startWith(first$),
                bufferCount(hotkeys.length, 1),
                timeoutWith(maxSequenceTimeout, output$)
              )
            ),
            matchEvents
          )

    const subscription = output$.subscribe(r => {
      handler(...(ensureArray(r) as KeyboardEvent[]))
    })
    return () => {
      scopedKeyBindings.delete(keys)
      if (!scopedKeyBindings.size) {
        keyBindings.delete(scopeId)
      }
      keyBindings$.next(new Map(keyBindings) as ScopedBindings)
      subscription.unsubscribe()
    }
  }
}

export const getKeyBindings = () => keyBindings
export const listenKeyBindings = (handler: (bindings: ScopedBindings) => void) => {
  const subscription = keyBindings$.subscribe(handler)
  return () => subscription.unsubscribe()
}

export const hotkey = createHotkey(canUseDOM ? document : undefined, {
  scope: 'global',
})


