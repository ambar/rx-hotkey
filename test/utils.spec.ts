import {matchObject, parseHotkeys} from '../src/utils'

describe('matchObject', () => {
  it('compares object', () => {
    expect(matchObject({foo: 1, bar: 2}, {foo: 1})).toBe(true)
    expect(matchObject({foo: 1}, {bar: 1})).toBe(false)
    expect(matchObject({foo: 1, bar: 2}, {foo: 1, baz: 2})).toBe(false)
  })

  it('uses custom comparator', () => {
    expect(matchObject({foo: 1, bar: 2}, {foo: 1}, Object.is)).toBe(true)
  })

  it.each(['ctrl+c', 'c', 'ctrl+alt+c', 'alt+c shift+c'])(
    'parses hotkey: %s',
    (key) => {
      expect(parseHotkeys(key)).toMatchSnapshot()
    }
  )
})
