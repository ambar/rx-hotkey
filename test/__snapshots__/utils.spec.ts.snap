// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchObject parses hotkey: alt+c shift+c 1`] = `
Array [
  Object {
    "altKey": true,
    "ctrlKey": false,
    "key": "c",
    "metaKey": false,
    "shiftKey": false,
  },
  Object {
    "altKey": false,
    "ctrlKey": false,
    "key": "c",
    "metaKey": false,
    "shiftKey": true,
  },
]
`;

exports[`matchObject parses hotkey: c 1`] = `
Array [
  Object {
    "altKey": false,
    "ctrlKey": false,
    "key": "c",
    "metaKey": false,
    "shiftKey": false,
  },
]
`;

exports[`matchObject parses hotkey: ctrl+alt+c 1`] = `
Array [
  Object {
    "altKey": true,
    "ctrlKey": true,
    "key": "c",
    "metaKey": false,
    "shiftKey": false,
  },
]
`;

exports[`matchObject parses hotkey: ctrl+c 1`] = `
Array [
  Object {
    "altKey": false,
    "ctrlKey": true,
    "key": "c",
    "metaKey": false,
    "shiftKey": false,
  },
]
`;
